syntax = "proto3"; // 声明了protobuf的版本

package fixbug; // 声明了代码所在的包（对于C++来说是namespace）

// 定义选项，标识生成 service 服务类，rpc 方法
// option 是 protoc 内置规定各种 option
option cc_generic_services = true;

// 自定义类型
// 1. 以下可以认为是远程调用的 RPC 所需的参数
message ResCode {
    int32 errcode = 1;
    bytes errmsg = 2;
}

message LoginRequest {
    bytes name = 1;
    bytes pwd  = 2;
}

message LoginResponse {
    ResCode code = 1;
    bool  success = 2;
}

message User {
    bytes name = 1;
    uint32 age = 2;
    enum Sex {
        MAN   = 0;
        WOMAN = 1;
    }
    Sex sex = 3;
}

message GetFriendListRequest {
    uint32 userid = 1;
}

// 应当返回自定义类型的列表
message GetFriendListResponse {
    ResCode code = 1;
    repeated User friend_list = 2;  // 表示元素为 User 类型的列表
}

// 2. 我们还需要定义各种 PRC 方法对应的类型

// UserServicePpc 也会成为一个类
service UserServicePpc {
    rpc Login(LoginRequest) returns(LoginResponse);  // 定义了一个 rpc 方法，指明接收实参及返回类型
    rpc GetFriendLists(GetFriendListRequest) returns(GetFriendListResponse);
}
